csi/controller_server.go

```diff
@@ -0,0 +1,269 @@
+		volSizeBytes = int64(req.GetCapacityRange().GetRequiredBytes())
+	}
+	volSizeGB := int(volSizeBytes / 1024 / 1024 / 1024)
+	vol.Size = fmt.Sprintf("%dGi", volSizeGB)
+
+	logrus.Infof("CreateVolume: creating a volume by API client, name: %s, size: %s", vol.Name, vol.Size)
+	resVol, err := cs.apiClient.Volume.Create(vol)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			Id:            resVol.Id,
+			CapacityBytes: int64(volSizeBytes),
+			Attributes:    req.GetParameters(),
+		},
+	}, nil
+}
+
+func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	logrus.Infof("ControllerServer delete volume req: %v", req)
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME); err != nil {
+		logrus.Errorf("DeleteVolume: invalid delete volume req: %v", req)
+		return nil, err
+	}
+
+	existVol, err := cs.apiClient.Volume.ById(req.GetVolumeId())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if existVol == nil {
+		msg := fmt.Sprintf("DeleteVolume: volume %s not exists", req.GetVolumeId())
+		logrus.Warn(msg)
+		return nil, status.Error(codes.Aborted, msg)
+	}
+
+	logrus.Debugf("DeleteVolume: volume %s exists", req.GetVolumeId())
+	if err = cs.apiClient.Volume.Delete(existVol); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.DeleteVolumeResponse{}, nil
+}
+
+func (cs *ControllerServer) ValidateVolumeCapabilities(ctx context.Context, req *csi.ValidateVolumeCapabilitiesRequest) (*csi.ValidateVolumeCapabilitiesResponse, error) {
+	logrus.Infof("ControllerServer ValidateVolumeCapabilities req: %v", req)
+	for _, cap := range req.GetVolumeCapabilities() {
+		if cap.GetAccessMode().GetMode() != csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER {
+			return &csi.ValidateVolumeCapabilitiesResponse{Supported: false, Message: ""}, nil
+		}
+	}
+	return &csi.ValidateVolumeCapabilitiesResponse{Supported: true, Message: ""}, nil
+}
+
+// ControllerPublishVolume will attach the volume to the specified node
+func (cs *ControllerServer) ControllerPublishVolume(ctx context.Context, req *csi.ControllerPublishVolumeRequest) (*csi.ControllerPublishVolumeResponse, error) {
+	logrus.Infof("ControllerServer ControllerPublishVolume req: %v", req)
+	existVol, err := cs.apiClient.Volume.ById(req.GetVolumeId())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if existVol == nil {
+		msg := fmt.Sprintf("ControllerPublishVolume: the volume %s not exists", req.GetVolumeId())
+		logrus.Warn(msg)
+		return nil, status.Error(codes.Aborted, msg)
+	}
+	if existVol.State == "attaching" {
+		return nil, status.Errorf(codes.Aborted, "The volume %s is attaching", req.GetVolumeId())
+	}
+
+	needToAttach := false
+	if existVol.State == "detached" {
```

Comment:
or `detaching`

Reply:
Check the volume state, return `Aborted` if `detaching`
```
if existVol.State == string(types.VolumeStateAttaching) || existVol.State == string(types.VolumeStateDetaching) {
    return nil, status.Errorf(codes.Aborted, "The volume %s is %s", req.GetVolumeId(), existVol.State)
}
```

Also return `Aborted` if attach timeout.